pkg/local_object_storage/metabase/revive.go

```diff
@@ -0,0 +1,155 @@
+package meta
+
+import (
+	"fmt"
+
+	"github.com/nspcc-dev/neofs-node/pkg/local_object_storage/util/logicerr"
+	"github.com/nspcc-dev/neofs-sdk-go/object"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"go.etcd.io/bbolt"
+)
+
+// ErrObjectWasNotRemoved is returned when object neither in the graveyard nor was marked with GC mark.
+var ErrObjectWasNotRemoved = logicerr.New("object neither in the graveyard nor was marked with GC mark")
+
+// ErrReviveFromContainerGarbage is returned when the object is in the container that marked with GC mark.
+var ErrReviveFromContainerGarbage = logicerr.New("revive from container marked with GC")
+
+type reviveStatusType int
+
+const (
+	// ReviveStatusGraveyard is the type of revival status of an object from a graveyard.
+	ReviveStatusGraveyard reviveStatusType = iota
+	// ReviveStatusGC is the type of revival status of an object from GC.
+	ReviveStatusGC
+	// ReviveStatusError is the type of status when an error occurs during revive.
+	ReviveStatusError
+)
+
+// ReviveStatus groups the resulting values of ReviveObject operation.
+// Contains the type of revival status and message for details.
+type ReviveStatus struct {
+	statusType reviveStatusType
+	message    string
+}
+
+// Message returns message of status.
+func (s *ReviveStatus) Message() string {
+	return s.message
+}
+
+// StatusType returns the type of revival status.
+func (s *ReviveStatus) StatusType() reviveStatusType {
+	return s.statusType
+}
+
+func (s *ReviveStatus) setStatusGraveyard(tomb string) {
+	s.statusType = ReviveStatusGraveyard
+	s.message = fmt.Sprintf("successful revival from graveyard, tomb: %s", tomb)
+}
+
+func (s *ReviveStatus) setStatusGC() {
+	s.statusType = ReviveStatusGC
+	s.message = "successful revive object with GC mark"
+}
+
+func (s *ReviveStatus) setStatusError(err error) {
+	s.statusType = ReviveStatusError
+	s.message = fmt.Sprintf("don't revive, err: %v", err)
+}
+
+// ReviveObject revives object by oid.Address. Removes GCMark/Tombstone records in the corresponding buckets
+// and restore metrics.
+func (db *DB) ReviveObject(addr oid.Address) (res ReviveStatus, err error) {
+	db.modeMtx.RLock()
+	defer db.modeMtx.RUnlock()
+
+	if db.mode.ReadOnly() {
+		res.setStatusError(ErrReadOnlyMode)
+		return res, ErrReadOnlyMode
+	} else if db.mode.NoMetabase() {
+		res.setStatusError(ErrDegradedMode)
+		return res, ErrDegradedMode
+	}
+
+	currEpoch := db.epochState.CurrentEpoch()
+
+	err = db.boltDB.Update(func(tx *bbolt.Tx) error {
+		garbageObjectsBKT := tx.Bucket(garbageObjectsBucketName)
+		garbageContainersBKT := tx.Bucket(garbageContainersBucketName)
+		graveyardBKT := tx.Bucket(graveyardBucketName)
+
+		buf := make([]byte, addressKeySize)
+
+		targetKey := addressKey(addr, buf)
+
+		if graveyardBKT == nil {
+			// incorrect metabase state, does not make
+			// sense to check garbage bucket
+			return ErrObjectWasNotRemoved
+		}
+
+		val := graveyardBKT.Get(targetKey)
+		if val != nil {
```

Comment:
if an object is in graveyard, do we skip garbage bucket then?

Reply:
After removal from the graveyard, the next operation I remove from the garbage bucket. But I noticed a duplicate code, so I've refractored it a little now, now after all the checks, anyway it removes it from the garbage bucket.